### PR TITLE
add reflexive method for Dimension

### LINKED ISSUE #17482: The "multiply" operator of type Dimension has a bug.
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Custom code
- **OS**: Win10 64bit 
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.5.0
- **Python version**: 3.6.2
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 9.0
- **GPU model and memory**: GTX 1080Ti
- **Exact command to reproduce**: See codes below

### Describe the problem
Suppose `A` is a tensor, `L = A.shape[0]`. I found that `L*3` works while `3*L` lead to a **TypeError**. Can someone fix this?

### Source code / logs
```
import tensorflow as tf
A = tf.placeholder( tf.float32, shape = [3, 4, 5] )
L = A.shape[0]
print( type(L) )
print( L*3 )
print( 3*L )
```
Output is:
image

### PR DESCRIPTION
**Pull request by facaiy**
Fix #17482

### CODE REVIEW

File: tensorflow/python/framework/tensor_shape_test.py [hunk 1]
```diff
@@ -168,6 +168,18 @@ def testInequality(self):
     self.assertIs(None, tensor_shape.Dimension(None) != None)  # pylint: disable=g-equals-none
     self.assertTrue(tensor_shape.Dimension(12) != 12.99)
 
+  def testCompatibleWithInt(self):
+    self.assertEqual(5 + tensor_shape.Dimension(3),
+                     tensor_shape.Dimension(5) + 3)
```
Comment (shoyer) [inline tensorflow/python/framework/tensor_shape_test.py]: These tests just verify that `int + Dimension` and `Dimension + int` give the same result, but they don't check the actual values.

Can you split these each of these asserts into two, comparing against explicitly constructed Dimension objects?

**Comment by shoyer** [comment]
This needs to be slightly more complicated because we can't safely cast every other type to a Dimension, only integers/None.

For example, currently Tensor + Dimension -> Tensor:
```
>>> tf.constant(1) + tf.Dimension(2)
<tf.Tensor 'add:0' shape=() dtype=int32>
```
This existing behavior would break with this patch.

On the other hand, Dimension + Tensor is currently an error. But it should probably return a Tensor instead (at least if the Dimension has known size).

To fix this, special methods like `__add__` and `__radd__` need to return `NotImplemented` when they encounter a type they don't recognize, instead of trying to cast everything to a Dimension. Something like:
```
def __add__(self, other):
  if other is not None and not isinstance(other, int):
    return NotImplemented
  # existing implementation
```

**Review by shoyer — CHANGES_REQUESTED** [review]

**Comment by facaiy** [comment]
@shoyer I think for Tensor + Dimension -> Tensor:
Tensor.__add__ is invoked before Dimension.__radd__, hence the patch won't break its behavior. I'd like to add more tests to confirm it.

Similarly, Dimension + Tensor, I think Dimension.__add__ won't work and so Tensor.__radd__ would be invoked.

File: tensorflow/python/framework/tensor_shape_test.py (hunk 1 above)
Comment (facaiy) [reply tensorflow/python/framework/tensor_shape_test.py]: Thanks for your suggestion. Yes, I'd like to add it later.

**Comment by facaiy** [comment]
@frankchn It seems that the build logs has been cleared. Could you restart all tests so I can check the log about failure on MacOS python2?

**Comment by facaiy** [comment]
I'll close the PR since 5fd341d3 has fixed it.

**Pull request closed without merging**